### PR TITLE
drivers: i2c: I2C shim for nrfx TWI and TWIM drivers

### PR DESCRIPTION
**Pull request by Mierunski**
I prepared shim implementation for TWI and TWIM nrfx drivers. Changed board definitions to point to the new shim and removed old implementation.
Tested with HTS221 example on nrf52_pca20020 board.

### CODE REVIEW

File: drivers/i2c/Kconfig.nrfx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+config I2C_0_NRF_TWIM
+	bool "nRF TWIM 0"
+	depends on SOC_SERIES_NRF52X
+	select NRFX_TWIM
+	help
+	  Enable nRF TWI Master with EasyDMA on port 0.
+
+endchoice
+
+config I2C_0_NRF_SDA_PIN
```
Comment (galak) [inline drivers/i2c/Kconfig.nrfx]: Can we move to getting these gpio's from device tree instead.

**Review by galak — CHANGES_REQUESTED** [review]
Since we are reworking the driver, lets use DTS for GPIOs related to SDA/SCL pins.

File: drivers/i2c/i2c_nrfx_twi.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		return -EBUSY;
+	}
+	nrfx_twi_enable(&get_dev_config(dev)->twi);
+	return 0;
+}
+
+#define I2C_NRFX_TWI_DEVICE(idx)					\
+	static int twi_##idx##_init(struct device *dev)			\
+	{								\
+		IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_TWI##idx),		\
```
Comment (anangl) [inline drivers/i2c/i2c_nrfx_twi.c]: Since we've got DTS entries for I2C, use `CONFIG_I2C_##idx##_IRQ` to get the IRQ number.

**Review by anangl — CHANGES_REQUESTED** [review]

File: drivers/i2c/Kconfig.nrfx (hunk 1 above)
Comment (Mierunski) [reply drivers/i2c/Kconfig.nrfx]: Moved pin settings to device tree, but also added an option to override those settings through kconfig - some DKs do not have specified pins and can be changed according to users needs.

File: drivers/i2c/i2c_nrfx_twi.c [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			.primary_length = msgs[i].len,
+			.address	= addr,
+			.type		= (msgs[i].flags & I2C_MSG_READ) ?
+					  NRFX_TWI_XFER_RX : NRFX_TWI_XFER_TX
+		};
+
+		nrfx_err_t res = nrfx_twi_xfer(&get_dev_config(dev)->twi,
+					       &cur_xfer,
+					       (msgs[i].flags & I2C_MSG_STOP) ?
+					       NRFX_TWI_FLAG_TX_NO_STOP : 0);
```
[the remark this answers is not in the document]
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twi.c]: Fixed

File: drivers/i2c/i2c_nrfx_twi.c [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+struct i2c_nrfx_twi_config {
+	nrfx_twi_t twi;
+};
+static inline struct i2c_nrfx_twi_data *get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+static inline const struct i2c_nrfx_twi_config
+*get_dev_config(struct device *dev)
```
[the remark this answers is not in the document]
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twi.c]: Fixed

File: drivers/i2c/i2c_nrfx_twi.c [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+{
+	nrfx_twi_t const *inst = &(get_dev_config(dev)->twi);
+
+	if (I2C_ADDR_10_BITS & dev_config) {
+		return -EINVAL;
+	}
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		inst->p_twi->FREQUENCY = NRF_TWI_FREQ_100K;
```
[the remark this answers is not in the document]
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twi.c]: Done

File: drivers/i2c/i2c_nrfx_twi.c (hunk 2 above)
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twi.c]: Done

File: drivers/i2c/i2c_nrfx_twi.c [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+					       &cur_xfer,
+					       (msgs[i].flags & I2C_MSG_STOP) ?
+					       0 : NRFX_TWI_FLAG_TX_NO_STOP);
+		if (res != NRFX_SUCCESS) {
+			return -EIO;
+		}
+
+		k_sem_take(&(get_dev_data(dev)->sync), K_FOREVER);
+		if (get_dev_data(dev)->res != NRFX_SUCCESS) {
+			return -EIO;
```
Comment (anangl) [inline drivers/i2c/i2c_nrfx_twi.c]: What about adding something like
`SYS_LOG_ERR("Error %d occurred for message %d", get_dev_data(dev)->res, i);`
one line above?

`res` could be copied to a local variable to avoid re-read (this field is volatile).

File: drivers/i2c/i2c_nrfx_twi.c [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+		dev_data->res = NRFX_ERROR_DRV_TWI_ERR_ANACK;
+		break;
+	case NRFX_TWI_EVT_DATA_NACK:
+		dev_data->res = NRFX_ERROR_DRV_TWI_ERR_DNACK;
+		break;
+	default:
+		dev_data->res = NRFX_ERROR_INTERNAL;
+		break;
+	}
+	k_sem_give(&(get_dev_data(dev)->sync));
```
Comment (anangl) [inline drivers/i2c/i2c_nrfx_twi.c]: `&dev_data->sync`

File: drivers/i2c/i2c_nrfx_twim.c [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+		dev_data->res = NRFX_ERROR_DRV_TWI_ERR_ANACK;
+		break;
+	case NRFX_TWIM_EVT_DATA_NACK:
+		dev_data->res = NRFX_ERROR_DRV_TWI_ERR_DNACK;
+		break;
+	default:
+		dev_data->res = NRFX_ERROR_INTERNAL;
+		break;
+	}
+	k_sem_give(&(get_dev_data(dev)->sync));
```
Comment (anangl) [inline drivers/i2c/i2c_nrfx_twim.c]: `&dev_data->sync`

File: drivers/i2c/i2c_nrfx_twim.c [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+	if (result != NRFX_SUCCESS) {
+		SYS_LOG_ERR("Failed to initialize device: %s",
+			    dev->config->name);
+		return -EBUSY;
+	}
+	nrfx_twim_enable(&get_dev_config(dev)->twim);
+	return 0;
+}
+
+#define I2C_NRFX_TWI_DEVICE(idx)					\
```
Comment (anangl) [inline drivers/i2c/i2c_nrfx_twim.c]: `I2C_NRFX_TWIM_DEVICE`

File: ext/hal/nordic/Kconfig [hunk 10]
```diff
@@ -21,3 +21,9 @@ config NRFX_SPIS
 
 config NRFX_WDT
 	bool
+
+config NRFX_TWI
+	bool
+
+config NRFX_TWIM
```
Comment (anangl) [inline ext/hal/nordic/Kconfig]: Place these entries before `NRFX_WDT` to keep the alphabetical order.

File: drivers/i2c/Kconfig.nrfx [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	select NRFX_TWI
+	help
+	  Enable nRF TWI Master without EasyDMA on port 0.
+
+config I2C_0_NRF_TWIM
+	bool "nRF TWIM 0"
+	depends on SOC_SERIES_NRF52X
+	select NRFX_TWIM
+	help
+	  Enable nRF TWI Master with EasyDMA on port 0.
```
Comment (anangl) [inline drivers/i2c/Kconfig.nrfx]: It would be good to mention here that this peripheral accepts only transfers from RAM, and when a buffer placed in flash is provided, the transfer will fail.

File: drivers/i2c/Kconfig.nrfx [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+	bool "nRF TWI nrfx drivers"
+	depends on SOC_FAMILY_NRF
+	help
+	  Enable support for nrfx TWI drivers for nRF MCU series.
+	  Peripherals with the same instance ID cannot be used together,
+	  e.g. I2C_0 and SPI_0. You may need to disable SPI_0 or SPI_1.
+
+if I2C_NRFX
+
+config I2C_NRFX_PINS_OVERRIDE_DTS
```
Comment (anangl) [inline drivers/i2c/Kconfig.nrfx]: Wouldn't `I2C_NRF_OVERRIDE_DTS_PINS` be more adequate (NRF, because this entry, like pins settings, is not nrfx related but generic for nRF chips)?

**Review by anangl — CHANGES_REQUESTED** [review]
A few more issues, mostly minor ones.

**Comment by kl-cruz** [comment]
I think this PR should trace changes in https://github.com/zephyrproject-rtos/zephyr/pull/7915/files#r192312703

File: boards/arm/bbc_microbit/Kconfig.defconfig [hunk 13]
```diff
@@ -28,22 +28,31 @@ endif # UART_NRF5
...
-endif # I2C
+endif #I2C_NRFX
 
-if I2C_NRF5
+if I2C_0
 
-config I2C_NRF5_0_GPIO_SCL_PIN
-	default 0
+choice I2C_0_NRF_TYPE
+	default I2C_0_NRF_TWIM
```
[the remark this answers is not in the document]
Comment (Mierunski) [reply boards/arm/bbc_microbit/Kconfig.defconfig]: Changed to TWI

File: drivers/i2c/i2c_nrfx_twi.c [hunk 14]
```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2017 - 2018, Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twi.c]: Fixed

File: drivers/i2c/i2c_nrfx_twi.c (hunk 6 above)
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twi.c]: Done

File: drivers/i2c/i2c_nrfx_twi.c (hunk 7 above)
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twi.c]: Fixed

File: drivers/i2c/i2c_nrfx_twim.c (hunk 8 above)
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twim.c]: Fixed

File: drivers/i2c/i2c_nrfx_twim.c (hunk 9 above)
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twim.c]: Fixed

File: drivers/i2c/Kconfig.nrfx (hunk 11 above)
Comment (Mierunski) [reply drivers/i2c/Kconfig.nrfx]: Done

File: drivers/i2c/Kconfig.nrfx (hunk 12 above)
Comment (Mierunski) [reply drivers/i2c/Kconfig.nrfx]: Changed

File: drivers/i2c/i2c_nrfx_twi.c [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+		dev_data->res = NRFX_ERROR_DRV_TWI_ERR_ANACK;
+		break;
+	case NRFX_TWI_EVT_DATA_NACK:
+		dev_data->res = NRFX_ERROR_DRV_TWI_ERR_DNACK;
+		break;
+	default:
+		dev_data->res = NRFX_ERROR_INTERNAL;
+		break;
+	}
+	k_sem_give(&dev_data->sync);
```
Comment (tbursztyka) [inline drivers/i2c/i2c_nrfx_twi.c]: add an empty line before

File: drivers/i2c/i2c_nrfx_twi.c [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+static int init_twi(struct device *dev, const nrfx_twi_config_t *config)
+{
+	nrfx_err_t result = nrfx_twi_init(&get_dev_config(dev)->twi, config,
+					  event_handler, dev);
+	if (result != NRFX_SUCCESS) {
+		SYS_LOG_ERR("Failed to initialize device: %s",
+			    dev->config->name);
+		return -EBUSY;
+	}
+	nrfx_twi_enable(&get_dev_config(dev)->twi);
```
Comment (tbursztyka) [inline drivers/i2c/i2c_nrfx_twi.c]: same

File: drivers/i2c/i2c_nrfx_twim.c [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+		}
+
+		k_sem_take(&(get_dev_data(dev)->sync), K_FOREVER);
+		res = get_dev_data(dev)->res;
+		if (res != NRFX_SUCCESS) {
+			SYS_LOG_ERR("Error %d occurred for message %d", res, i);
+			return -EIO;
+		}
+	}
+	return 0;
```
Comment (tbursztyka) [inline drivers/i2c/i2c_nrfx_twim.c]: same

File: drivers/i2c/i2c_nrfx_twim.c [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+		dev_data->res = NRFX_ERROR_DRV_TWI_ERR_ANACK;
+		break;
+	case NRFX_TWIM_EVT_DATA_NACK:
+		dev_data->res = NRFX_ERROR_DRV_TWI_ERR_DNACK;
+		break;
+	default:
+		dev_data->res = NRFX_ERROR_INTERNAL;
+		break;
+	}
+	k_sem_give(&dev_data->sync);
```
Comment (tbursztyka) [inline drivers/i2c/i2c_nrfx_twim.c]: etc...

File: drivers/i2c/i2c_nrfx_twi.c (hunk 15 above)
Comment (Mierunski) [reply drivers/i2c/i2c_nrfx_twi.c]: done

File: boards/arm/nrf52_pca20020/board.c [hunk 19]
```diff
@@ -12,6 +12,8 @@
 #define VDD_PWR_CTRL_GPIO_PIN 30
 #define CCS_VDD_PWR_CTRL_GPIO_PIN 10
 
+#ifdef CONFIG_GPIO
+
```
Comment (anangl) [inline boards/arm/nrf52_pca20020/board.c]: This change does not seem to be related to TWI/TWIM shims or I2C DTS entries, so it should not be part of this PR.
And is it needed at all?

File: drivers/i2c/Kconfig.nrfx [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+	select NRFX_TWI
+	help
+	  Enable nRF TWI Master without EasyDMA on port 1.
+
+config I2C_1_NRF_TWIM
+	bool "nRF TWIM 1"
+	depends on SOC_SERIES_NRF52X
+	select NRFX_TWIM
+	help
+	  Enable nRF TWI Master with EasyDMA on port 1.
```
Comment (anangl) [inline drivers/i2c/Kconfig.nrfx]: Add the same note regarding transfers from RAM as for I2C_0_NRF_TWIM.

**Review by anangl — CHANGES_REQUESTED** [review]
Change the title of the 4th commit to "boards: nrf: Moved SDA and SCL pin configuration to DTS".
And in 1st commit "shim" -> "shims" (but it's just a detail).

File: boards/arm/nrf52_pca20020/board.c (hunk 19 above)
Comment (anangl) [reply boards/arm/nrf52_pca20020/board.c]: On second thought, I think this issue should be solved in a different way. The GPIO module should be explicitly enabled in the default configuration for pca20020 (previously, it was enabled implicitly by the old I2C driver), so that the code in `board.c` could turn on the VDD power rail. And this power rail is needed for example for the SX1509B GPIO expander to operate, and this expander is enabled by default for this board (and it depends on GPIO as well). Thus, I think the assumption of the author of this part of code was that GPIO should be always enabled (hence, CONFIG_GPIO was not checked). @tkln Could you confirm my understanding is correct?

**Review by anangl — CHANGES_REQUESTED** [review]

File: boards/arm/nrf52_pca20020/board.c (hunk 19 above)
Comment (Mierunski) [reply boards/arm/nrf52_pca20020/board.c]: Done

File: drivers/i2c/Kconfig.nrfx (hunk 20 above)
Comment (Mierunski) [reply drivers/i2c/Kconfig.nrfx]: Done

**Review by anangl — APPROVED** [review]

**Comment by carlescufi** [comment]
@galak can you revisit this now that the release is out of the way?

**Review by carlescufi — APPROVED** [review]
LGTM now

File: drivers/i2c/Kconfig.nrfx [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+	bool "nRF TWI nrfx drivers"
+	depends on SOC_FAMILY_NRF
+	help
+	  Enable support for nrfx TWI drivers for nRF MCU series.
+	  Peripherals with the same instance ID cannot be used together,
+	  e.g. I2C_0 and SPI_0. You may need to disable SPI_0 or SPI_1.
+
+if I2C_NRFX
+
+config I2C_NRF_OVERRIDE_DTS_PINS
```
Comment (galak) [inline drivers/i2c/Kconfig.nrfx]: Why do we need this?

**Review by galak — DISMISSED** [review]

File: drivers/i2c/Kconfig.nrfx (hunk 21 above)
Comment (Mierunski) [reply drivers/i2c/Kconfig.nrfx]: It is mainly used for development kits, where I2C pins can be changed quite often, depending on hardware connected to the board.
We can't use configuration from DTS as default pins for Kconfig, but by adding override option we can switch to using pin configuration from DTS or to using user configuration from menuconfig.

Comment (galak) [reply drivers/i2c/Kconfig.nrfx]: > It is mainly used for development kits, where I2C pins can be changed quite often, depending on hardware connected to the board.
> We can't use configuration from DTS as default pins for Kconfig, but by adding override option we can switch to using pin configuration from DTS or to using user configuration from menuconfig.

I'm not keen on having support for this, we should be doing HW cfg via the DTS and only DTS.

Comment (carlescufi) [reply drivers/i2c/Kconfig.nrfx]: Actually since there is an active discussion regarding Kconfig and DT I'll copy a couple of people here who have been involved so we get more perspectives: @erwango @tbursztyka @SebastianBoe

Comment (erwango) [reply drivers/i2c/Kconfig.nrfx]: think having multiple ways for HW configuration (or anything in general) is not a good idea.
Some will use one way, some will use other, it'll soon become difficult to maintain.

Comment (tbursztyka) [reply drivers/i2c/Kconfig.nrfx]: Aren't sda/scd pins the same on 1 board, no matter how many slaves you wire up?

Comment (jarz-nordic) [reply drivers/i2c/Kconfig.nrfx]: > Aren't sda/scd pins the same on 1 board, no matter how many slaves you wire up?

No, in NRF SoCs you can connect SDA/SDC PINs wherevere you wish and you can modify them during runtime.
I saw that for instance ST can have 2 sets of PINs for UART in DTS. We can be more flexible thus we would like to have  possibility to overwride DTS in Kconfig. Putting all possible combinations to DTS makes limited sense in my opinion.

**Comment by carlescufi** [comment]
@galak your request was addressed from what I can tell, can your revisit?
